trunner/config.py

```diff
@@ -25,145 +40,255 @@
-    """Exception raised when config has wrong format"""
+class ParserError(Exception):
+    pass
+
+
+def array_value(array: Dict[str, List[str]]) -> List[str]:
+    value = set(array.get('value', []))
+    value |= set(array.get('include', []))
+    value -= set(array.get('exclude', []))
+
+    return list(value)
+
 
-    def __init__(self, parser, message="Parser error"):
-        message = f"{parser.path}:\n{message}"
-        super().__init__(message)
+@dataclass
+class ParserArgs:
+    targets: List[str]
+    yaml_path: Path
+    path: Path = field(init=False)
 
+    def __post_init__(self):
+        self.path = self.yaml_path.parent
 
-class YAMLParser:
-    """This class is responsible for parsing YAML config that defines test cases"""
 
-    def __init__(self, path=None, targets=None):
-        self.config = None
-        if targets is None:
-            targets = ALL_TARGETS
-        self.targets = targets
-        self.parsed_tests = []
+class Config(dict):
+    def __init__(self, config: dict) -> None:
+        self.update(**config)
 
-        if path is None:
-            self.path = ''
+    @classmethod
+    def from_dict(cls, config: dict) -> 'Config':
+        parser = ConfigParser()
+        instance = cls(config)
+        parser.parse(instance)
+        return instance
+
+    def join_targets(self, config: 'Config') -> None:
+        if 'targets' not in config:
+            return
+        elif 'targets' not in self:
+            self['targets'] = config['targets']
             return
 
-        self.path = path
-        self.load_yaml(self.path)
-        # Current path is /example/test/test.yaml - get rid of test.yaml in path
-        self.path = self.path.parents[0]
+        if 'value' in self['targets']:
+            # The value field is defined. Do not overwrite it, just return.
+            return
 
-    def load_yaml(self, path):
-        with open(path, 'r') as f_yaml:
+        targets = dict(config['targets'])
+        targets.setdefault('value', ALL_TARGETS)
+        value = array_value(targets)
+        self['targets']['value'] = value
+
+    def join(self, config: 'Config') -> None:
+        self.join_targets(config)
+        for key in config:
+            if key not in self:
+                self[key] = config[key]
+
+    def setdefault_targets(self) -> None:
+        targets = self.get('targets', dict())
+        targets.setdefault('value', DEFAULT_TARGETS)
+        self['targets'] = targets
+
+    def setdefaults(self) -> None:
+        self.setdefault('ignore', False)
+        self.setdefault('type', 'unit')
+        self.setdefault('timeout', PYEXPECT_TIMEOUT)
+        self.setdefault_targets()
+
+
+class TestConfig(Config):
+    def __init__(self, config: dict) -> None:
+        super().__init__(config)
+
+    @classmethod
+    def from_dict(cls, config: dict, main: Config, args: ParserArgs) -> 'TestConfig':
+        parser = ConfigParser()
+        instance = cls(config)
+        parser.parse(instance)
```

Comment:
We could use `super` but I have little problem with typehints, here some generic type must be used I bet. Lets leave it as is and I will probably change it with adding mypy to CI. This will also fix mypy warnings.